src/target/riscv/riscv-013.c

```diff
@@ -3043,3 +3045,472 @@ void riscv013_clear_abstract_error(struct target *target)
+		}
+	}
+
+	/* Single-Step each hart. */
+	for (int hartsel = 0; hartsel < riscv_count_harts(target); hartsel++) {
+		riscv_set_current_hartid(target, hartsel);
+		riscv013_on_step(target);
+		riscv013_step_current_hart(target);
+		COMPLIANCE_TEST(riscv_halt_reason(target, hartsel) == RISCV_HALT_SINGLESTEP,
+				"Single Step should result in SINGLESTEP");
+	}
+
+	/* Core Register Tests */
+	uint64_t bogus_dpc = 0xdeadbeef;
+	for (int hartsel = 0; hartsel < riscv_count_harts(target); hartsel++) {
+		riscv_set_current_hartid(target, hartsel);
+
+		/* DCSR Tests */
+		register_write_direct(target, GDB_REGNO_DCSR, 0x0);
+		register_read_direct(target, &value, GDB_REGNO_DCSR);
+		COMPLIANCE_TEST(value != 0,	"Not all bits in DCSR are writable by Debugger");
+		register_write_direct(target, GDB_REGNO_DCSR, 0xFFFFFFFF);
+		register_read_direct(target, &value, GDB_REGNO_DCSR);
+		COMPLIANCE_TEST(value != 0,	"At least some bits in DCSR must be 1");
+
+		/* DPC. Note that DPC is sign-extended. */
+		riscv_reg_t dpcmask = 0xFFFFFFFCUL;
+		riscv_reg_t dpc;
+
+		if (riscv_xlen(target) > 32)
+			dpcmask |= (0xFFFFFFFFULL << 32);
+
+		if (riscv_supports_extension(target, riscv_current_hartid(target), 'C'))
+			dpcmask |= 0x2;
+
+		register_write_direct(target, GDB_REGNO_DPC, dpcmask);
+		register_read_direct(target, &dpc, GDB_REGNO_DPC);
+		COMPLIANCE_TEST(dpcmask == dpc,
+				"DPC must be sign-extended to XLEN and writable to all-1s (except the least significant bits)");
+		register_write_direct(target, GDB_REGNO_DPC, 0);
+		register_read_direct(target, &dpc, GDB_REGNO_DPC);
+		COMPLIANCE_TEST(dpc == 0, "DPC must be writable to 0.");
+		if (hartsel == 0)
+			bogus_dpc = dpc; /* For a later test step */
+	}
+
+	/* NDMRESET
+	Asserting non-debug module reset should not reset Debug Module state.
+	But it should reset Hart State, e.g. DPC should get a different value.
+	Also make sure that DCSR reports cause of 'HALT' even though previously we single-stepped.
+	*/
+
+	/* Write some registers. They should not be impacted by ndmreset. */
+	dmi_write(target, DMI_COMMAND, 0xFFFFFFFF);
+
+	for (unsigned int i = 0; i < get_field(abstractcs, DMI_ABSTRACTCS_PROGBUFSIZE); i++) {
+		testvar = (i + 1) * 0x11111111;
+		dmi_write(target, DMI_PROGBUF0 + i, testvar);
+	}
+
+	for (unsigned int i = 0; i < get_field(abstractcs, DMI_ABSTRACTCS_DATACOUNT); i++) {
+		testvar = (i + 1) * 0x11111111;
+		dmi_write(target, DMI_DATA0 + i, testvar);
+	}
+
+	dmi_write(target, DMI_ABSTRACTAUTO, 0xFFFFFFFF);
+	dmi_read(target, &abstractauto, DMI_ABSTRACTAUTO);
+
+	/* Pulse reset. */
+	target->reset_halt = true;
+	riscv_set_current_hartid(target, 0);
+	COMPLIANCE_TEST(ERROR_OK == assert_reset(target), "Must be able to assert NDMRESET");
+	COMPLIANCE_TEST(ERROR_OK == deassert_reset(target), "Must be able to deassert NDMRESET");
+
+	/* Verify that most stuff is not affected by ndmreset. */
+	dmi_read(target, &testvar_read, DMI_COMMAND);
```

Comment:
command is not readable.

Reply:
right, that was clarified later, so I'll update the test.